### PR TITLE
Lufactmodn

### PR DESCRIPTION
**Pull request by arghhhh**
Issue JuliaLang/LinearAlgebra.jl#431
This is my attempt to add the third option between no pivoting and pivoting using the absmax element.  It obviously breaks the current API, but all `linalg` tests pass.
I have replaced:
`Pivot = Val{true}` with `Pivot = Val{:maxabspivot}`
`Pivot = Val{false}` with `Pivot = Val{:nopivot}`
and added the option `Pivot = Val{:nonzeropivot}`
This now allows A\b without having to define hacky and unwanted versions of `abs` and `<`.

### CODE REVIEW

**Comment by andreasnoack** [comment]
Great. Thanks for contributing this. The functionality is needed. I'm not sure about the names yet. What would people think about pivoting types instead of `Val`s? When we have decided, we should deprecate the old behavior and also change `qr` such that all pivoting remains consistent.

**Comment by arghhhh** [comment]
It'd be good to be able to abstract out the concept of pivoting method as a function of element type.
I had to add the line:
`Base.lufact{N}(A::AbstractMatrix{ModInt{N}}, pivot::Union{Type{Val{:maxabspivot}}, Type{Val{:nopivot}}, Type{Val{:nonzeropivot}} } = Val{:nonzeropivot}) = lufact!(copy(A), pivot)`
in order to get `A\b` to work.
I think we need some sort of trait function that would map ModInt{N} to `:nonzeropivot`, so that authors of new types can override the default pivoting policy cleanly, without the line above. 
Any suggestions on this?

**Comment by kshyatt** [comment]
What's the status on this?

**Comment by andreasnoack** [comment]
The status right now is JuliaCon. I might have time to review during the hackathon. If not, I'll take a look next week.

File: test/linalg/generic.jl [hunk 1]
```diff
@@ -332,13 +332,22 @@ Base.one(::Type{ModInt{n}}) where {n} = ModInt{n}(1)
...
-@test A*(lufact(A, Val{false})\b) == b
+@test A*(lufact(A, Val{:nopivot})\b) == b
 
-# Needed for pivoting:
+# nonzero pivoting required:
+A = [ModInt{2}(0) ModInt{2}(1); ModInt{2}(1) ModInt{2}(1)]
+
+@test A*(lufact(A, Val{:nonzeropivot})\b) == b
+
+Base.lufact{N}(A::AbstractMatrix{ModInt{N}}, pivot::Union{Type{Val{:maxabspivot}}, Type{Val{:nopivot}}, Type{Val{:nonzeropivot}} } = Val{:nonzeropivot}) = lufact!(copy(A), pivot)
```
Comment (tkelman) [inline test/linalg/generic.jl]: should wrap long lines

**Comment by ViralBShah** [comment]
Closing as stale. Probably needs to be a new PR anyways.

**Pull request closed without merging**